llama.cpp

```diff
@@ -17628,7 +17628,7 @@ static int32_t llama_chat_apply_template_internal(
             }
         }
         // llama2 templates seem to not care about "add_generation_prompt"
-    } else if (tmpl == "zephyr" || tmpl.find("<|user|>") != std::string::npos) {
+    } else if (tmpl == "zephyr" || (tmpl.find("<|user|>") != std::string::npos && tmpl.find("<|end|>") == std::string::npos)) {
```

Comment:
I could not easily confirm the chat template for zephyr, a random GGUF of a random finetune had this in the GGUF metadata:
```
{% for message in messages %} {% if message['role'] == 'user' %} {{ '<|user|> ' + message['content'] + eos_token }} {% elif message['role'] == 'system' %} {{ '<|system|> ' + message['content'] + eos_token }} {% elif message['role'] == 'assistant' %} {{ '<|assistant|> ' + message['content'] + eos_token }} {% endif %} {% if loop.last and add_generation_prompt %} {{ '<|assistant|>' }} {% endif %} {% endfor %}
```

So can not match on `<|endoftext|>` As it doesn't appear explicitly. `eos_token` is also not an option is it appears in both the zephyr & old phi3 template. 
Therefore I decided to falsify the zephyr match when <|end|> is in the template, which is the case for both phi3 templates.

Reply:
I'd recommend to move the phi-3 check before the check for zephyr, instead of doing inverse logic like this.

Example:

```
else if (tmpl == "phi3" || (tmpl.find("<|user|>") != std::string::npos && tmpl.find("<|end|>") != std::string::npos)) {
   ...
} else if (tmpl == "zephyr" || tmpl.find("<|user|>") != std::string::npos) {
   ...
} else if ...
```

Also, phi-3 templates seems to have the distinct `+ ' '` condition. Why don't we do: `tmpl.find("<|user|>") != std::string::npos && tmpl.find("+ ' '") != std::string::npos)` ?

Reply:
I reordered & removed the inverse check like suggested.
I also replaced the `find("<|end|>")` with `find("+ ' '")`.

`tmpl.find("<|user|>")` wouldn't work for the old phi3 template which uses `'<|' + message['role'] + '|>'` instead of `<|user|>`

Reply:
According to discussion from https://github.com/ggerganov/llama.cpp/pull/6857/files/992fff725f307ce54e02586f82a0303e6b0023b6#diff-b87e1c2fc779778832f03ef373963fcc40603a0500fc13cdc4cc570ee0189e8f , the given template may not be correct. Can you have a link to the original `tokenizer_config.json` file?

Reply:
I changed the match to find("<|end|>") again, I'm also fine with find("+ '\n'") if you think that's better...

Reply:
We should `find("+ '\n'")` because this condition appears in almost all templates - not just phi-3.

The reason why I suggested `find("+ ' '")` was because I thought that phi-3 uses space after special token - which turns out to be incorrect.